Mathlib/LinearAlgebra/Eigenspace/Basic.lean

```diff
@@ -58,227 +58,549 @@ open FiniteDimensional Set
 variable {K R : Type v} {V M : Type w} [CommRing R] [AddCommGroup M] [Module R M] [Field K]
   [AddCommGroup V] [Module K V]
 
+lemma ker_pow_le_ker_pow_of_le (f : End R M) (k l : ℕ) (hkl : k ≤ l) :
+    LinearMap.ker (f ^ k) ≤ LinearMap.ker (f ^ l) := by
+  rw [← pow_sub_mul_pow _ hkl]
+  exact LinearMap.ker_le_ker_comp _ _
+
+/-- The submodule `unifEigenspace f μ k` for a linear map `f`, a scalar `μ`,
+and a number `k : ENat` is the kernel of `(f - μ • id) ^ k` if `k` is a natural number,
+or the union of all these kernels if `k = ∞`. -/
+def unifEigenspace (f : End R M) (μ : R) (k : ENat) : Submodule R M :=
+  ⨆ l : ℕ, ⨆ _ : l ≤ k, LinearMap.ker ((f - μ • 1) ^ l)
+
+lemma mem_unifEigenspace {f : End R M} {μ : R} {k : ENat} {x : M} :
+    x ∈ f.unifEigenspace μ k ↔ ∃ l : ℕ, l ≤ k ∧ x ∈ LinearMap.ker ((f - μ • 1) ^ l) := by
+  have : Nonempty {l : ℕ // l ≤ k} := ⟨⟨0, zero_le _⟩⟩
+  rw [unifEigenspace, iSup_subtype', Submodule.mem_iSup_of_directed]
+  simp only [LinearMap.mem_ker, Subtype.exists, exists_prop]
+  intro m n
+  use max m n
+  constructor <;> apply ker_pow_le_ker_pow_of_le <;> simp
+
+lemma unifEigenspace_mono (f : End R M) (μ : R) : Monotone (f.unifEigenspace μ) := by
+  intro k l hkl x
+  rw [mem_unifEigenspace, mem_unifEigenspace]
+  rintro ⟨n, hn, hx⟩
+  use n, hn.trans hkl, hx
+
+lemma unifEigenspace_directed {f : End R M} {μ : R} {k : ℕ∞} :
+    Directed (· ≤ ·) (fun l : {l : ℕ // l ≤ k} ↦ f.unifEigenspace μ l) := by
+  have aux : Monotone ((↑) : {l : ℕ // l ≤ k} → ℕ∞) := fun x y h ↦ by simpa using h
+  exact ((unifEigenspace_mono f μ).comp aux).directed_le
+
+lemma mem_unifEigenspace_nat {f : End R M} {μ : R} {k : ℕ} {x : M} :
+    x ∈ f.unifEigenspace μ k ↔ x ∈ LinearMap.ker ((f - μ • 1) ^ k) := by
+  rw [mem_unifEigenspace]
+  constructor
+  · rintro ⟨l, hl, hx⟩
+    simp only [Nat.cast_le] at hl
+    apply ker_pow_le_ker_pow_of_le _ _ _ hl hx
+  · intro hx
+    exact ⟨k, le_rfl, hx⟩
+
+lemma mem_unifEigenspace_top {f : End R M} {μ : R} {x : M} :
+    x ∈ f.unifEigenspace μ ⊤ ↔ ∃ k : ℕ, x ∈ LinearMap.ker ((f - μ • 1) ^ k) := by
+  simp [mem_unifEigenspace]
+
+lemma unifEigenspace_nat {f : End R M} {μ : R} {k : ℕ} :
+    f.unifEigenspace μ k = LinearMap.ker ((f - μ • 1) ^ k) := by
+  ext; simp [mem_unifEigenspace_nat]
+
+lemma unifEigenspace_eq_iSup_unifEigenspace_nat (f : End R M) (μ : R) (k : ENat) :
+    f.unifEigenspace μ k = ⨆ l : {l : ℕ // l ≤ k}, f.unifEigenspace μ l := by
+  rw [unifEigenspace]
+  simp only [iSup_subtype, unifEigenspace_nat]
+
+lemma unifEigenspace_one {f : End R M} {μ : R} :
+    f.unifEigenspace μ 1 = LinearMap.ker (f - μ • 1) := by
+  rw [← Nat.cast_one, unifEigenspace_nat, pow_one]
+
+lemma mem_unifEigenspace_one {f : End R M} {μ : R} {x : M} :
+    x ∈ f.unifEigenspace μ 1 ↔ f x = μ • x := by
+  rw [unifEigenspace_one, LinearMap.mem_ker, LinearMap.sub_apply,
+    sub_eq_zero, LinearMap.smul_apply, LinearMap.one_apply]
+
+lemma mem_unifEigenspace_zero {f : End R M} {μ : R} {x : M} :
```

Comment:
And this?
```suggestion
@[simp] lemma mem_unifEigenspace_zero {f : End R M} {μ : R} {x : M} :
```

Reply:
simp can prove this one using the lemma below it.